### PR TITLE
Use :sets v2 under the hood in MapSet

### PR DESCRIPTION
**Pull request by whatyouhide**
We can do this now that Elixir requires OTP 24+, since the new :sets
API is also built on top of maps.

Part of https://github.com/elixir-lang/elixir/issues/10433.

### CODE REVIEW

File: lib/elixir/lib/map_set.ex [hunk 1]
```diff
@@ -177,35 +154,12 @@ defmodule MapSet do
...
-      iter
-      |> :maps.next()
-      |> disjointer(Map.delete(small, key), list)
-    else
-      iter
-      |> :maps.next()
-      |> disjointer(small, [key | list])
-    end
+  def symmetric_difference(%MapSet{set: set1} = map_set1, %MapSet{set: set2} = _map_set2) do
+    %{map_set1 | set: :sets.union(:sets.subtract(set1, set2), :sets.subtract(set2, set1))}
```
Comment (whatyouhide) [inline lib/elixir/lib/map_set.ex:162]: This is the only part I’m thinking about: is this ok? Or should we use `sets:fold/3` to do this more similarly to what we were doing before?

File: lib/elixir/lib/map_set.ex [hunk 2]
```diff
@@ -409,21 +332,9 @@ defmodule MapSet do
...
-
-  defp filter_keys({key, _value, iter}, fun) do
-    if fun.(key) do
-      [key | filter_keys(:maps.next(iter), fun)]
-    else
-      filter_keys(:maps.next(iter), fun)
-    end
+  def filter(%MapSet{set: set} = map_set, fun) when is_function(fun) do
+    pred = fn element -> !!fun.(element) end
+    %{map_set | set: :sets.filter(pred, set)}
```
Comment (whatyouhide) [inline lib/elixir/lib/map_set.ex:337]: Is it ok to use `:sets.filter/2` with an on-the-fly predicate, or should also go with `:sets.fold/3` here?

Comment (josevalim) [reply lib/elixir/lib/map_set.ex:337]: I would also keep it as it was, even though it is opaque.

Comment (josevalim) [reply lib/elixir/lib/map_set.ex:337]: I think the new version is fine. :)

File: lib/elixir/lib/map_set.ex (hunk 1 above)
Comment (josevalim) [reply lib/elixir/lib/map_set.ex:162]: I think we should use sets fold, yes. It should be more efficient.

Comment (whatyouhide) [reply lib/elixir/lib/map_set.ex]: @josevalim cool, updated!

**Comment by lukaszsamson** [comment]
This change will break backwards compatibility when elixir code compiled on <1.15 is run on >=1.15 and when elixir terms are sent over the wire when nodes running on different versions are connected in a cluster. See https://github.com/elixir-lang/elixir/issues/11110

**Comment by josevalim** [comment]
@lukaszsamson the representation of sets v2 is exactly the same representation as we had. :)

File: lib/elixir/lib/map_set.ex [hunk 3]
```diff
@@ -177,35 +154,23 @@ defmodule MapSet do
...
-      |> :maps.next()
-      |> disjointer(small, [key | list])
+  def symmetric_difference(%MapSet{set: set1} = map_set1, %MapSet{set: set2} = _map_set2) do
+    {small, large} = if :sets.size(set1) <= :sets.size(set2), do: {set1, set2}, else: {set2, set1}
+
+    disjointer_fun = fn elem, {small, acc} ->
+      if :sets.is_element(elem, small) do
+        {:sets.del_element(elem, small), acc}
+      else
+        {small, :sets.add_element(elem, acc)}
```
Comment (josevalim) [inline lib/elixir/lib/map_set.ex:168]: Don't do `add_element` here. Start with an empty list, collect keys, and then convert it to sets with `from_list` at the end. It is much more efficient.

**Review by josevalim — APPROVED** [review]
Just one comment left and ship it!

**Comment by lukaszsamson** [comment]
> @lukaszsamson the representation of sets v2 is exactly the same representation as we had. :)

but the implementation now pattern matches on map key `set` that is not present in terms originating from older versions.

File: lib/elixir/lib/map_set.ex (hunk 3 above)
Comment (whatyouhide) [reply lib/elixir/lib/map_set.ex:168]: @josevalim ah good callout. Pushed!

**Comment by josevalim** [comment]
@lukaszsamson excellent call!

@whatyouhide we need to keep the internal map key the same. We cannot rename it from `map` to `set`. It must remain `map`. You can add a comment explaining that sets v2 are just maps.

**Comment by whatyouhide** [comment]
@lukaszsamson great call 🙇 @josevalim updated once more. 🙃

**Pull request merged**